### PR TITLE
feat: Support all java.lang.Math in the query language

### LINKED ISSUE #5995: Support all `java.lang.Math` in the query language
All [`java.lang.Math`](https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html) methods should be supported in the query language.

Methods known to be missing are:

- [ ] addExact
- [ ] atan2
- [ ] cbrt
- [ ] copySign
- [ ] decrementExact
- [ ] expm2
- [ ] floorDiv
- [ ] floorMod
- [ ] getExponent
- [ ] hypot
- [ ] IEEEremainder
- [ ] incrementExact
- [ ] log10
- [ ] log1p
- [ ] multiplyExact
- [ ] negateExact
- [ ] nextAfter
- [ ] nextDown
- [ ] nextUp
- [ ] scalb
- [ ] sinh
- [ ] subtractExact
- [ ] tanh
- [ ] toDegrees
- [ ] toIntExact
- [ ] toRadians
- [ ] ulp
- [ ] E (constant)
- [ ] PI (constant)

### PR DESCRIPTION
**Pull request by chipkent**
Newly supported methods and constants:

- [ ] addExact
- [ ] atan2
- [ ] cbrt
- [ ] copySign
- [ ] cosh
- [ ] decrementExact
- [ ] expm1
- [ ] floorDiv
- [ ] floorMod
- [ ] getExponent
- [ ] hypot
- [ ] IEEEremainder
- [ ] incrementExact
- [ ] log10
- [ ] log1p
- [ ] multiplyExact
- [ ] negateExact
- [ ] nextAfter
- [ ] nextDown
- [ ] nextUp
- [ ] scalb
- [ ] sinh
- [ ] subtractExact
- [ ] tanh
- [ ] toDegrees
- [ ] toRadians
- [ ] toIntExact
- [ ] toShortExact
- [ ] toByteExact
- [ ] ulp
- [ ] E (constant)
- [ ] PI (constant)

Resolves #5995.

### CODE REVIEW

File: engine/function/src/templates/Numeric.ftl [hunk 1]
```diff
@@ -3378,6 +3390,600 @@ public class Numeric {
...
+     */
+    static public int toIntExact(${pt.primitive} x) {
+        if (isNull(x)) {
+            return NULL_INT;
+        }
+
+        <#if pt.primitive == "byte" || pt.primitive == "short" || pt.primitive == "int" >
+        return x;
+        <#else>
+        return Math.toIntExact(x);
```
Comment (kosak) [inline engine/function/src/templates/Numeric.ftl:3914]: This is a weird case but bear with me.
I don't like what happens when pt.primitive is `long` and the long value happens to be `Integer.MIN_VALUE`

This is a perfectly valid long value and in particular is not Deephaven NULL.

However, in this case this method would return Integer.MIN_VALUE, i.e. a deephaven NULL.

So in summary it took a long value that was *not* a Deephaven NULL and turned it into an integer value that *is* a Deephaven NULL.

I think I would test the result of Math.toIntExact with IsNull and throw an ArtihmeticException if it returns true.  That is, this case should be treated like an overflow.

File: engine/function/src/templates/Numeric.ftl [hunk 2]
```diff
@@ -3378,6 +3390,600 @@ public class Numeric {
...
+        }
+
+        <#if pt.primitive == "byte" || pt.primitive == "short" >
+        return x;
+        <#else>
+        if (x > Short.MAX_VALUE || x < Short.MIN_VALUE) {
+            throw new ArithmeticException("Overflow: ${pt.primitive} value will not fit in a short" + x);
+        }
+
+        return (short) x;
```
Comment (kosak) [inline engine/function/src/templates/Numeric.ftl:3939]: Similar to the above, this has a Deephaven NULL problem if the argument type is int or long and the value happens to be `Short.MIN_VALUE`

To reuse the same exception, I'd suggest writing it like

```
        short result = (short)x;
        if (x > Short.MAX_VALUE || x < Short.MIN_VALUE || isNull(result)) {
            throw new ArithmeticException("Overflow: ${pt.primitive} value will not fit in a short " + x);
        }
       return result;
```

and notice the extra space after "fit in a short"

File: engine/function/src/templates/Numeric.ftl [hunk 3]
```diff
@@ -3378,6 +3390,600 @@ public class Numeric {
...
+        }
+
+        <#if pt.primitive == "byte" >
+        return x;
+        <#else>
+        if (x > Byte.MAX_VALUE || x < Byte.MIN_VALUE) {
+            throw new ArithmeticException("Overflow: ${pt.primitive} value will not fit in a byte" + x);
+        }
+
+        return (byte) x;
```
Comment (kosak) [inline engine/function/src/templates/Numeric.ftl:3964]: Same as the above, so

```
        byte result = (byte)x;
        if (x > Byte.MAX_VALUE || x < Byte.MIN_VALUE || isNull(result)) {
            throw new ArithmeticException("Overflow: ${pt.primitive} value will not fit in a byte " + x);
        }
       return result;
```

File: engine/function/src/templates/Numeric.ftl [hunk 4]
```diff
@@ -23,6 +24,17 @@ import static io.deephaven.function.Cast.castDouble;
 @SuppressWarnings({"RedundantCast", "unused", "ManualMinMaxCalculation"})
 public class Numeric {
 
+    //////////////////////////// Constants ////////////////////////////
+
+    /**
+     * The double value that is closer than any other to e, the base of the natural logarithms.
```
[the remark this answers is not in the document]
Comment (chipkent) [reply engine/function/src/templates/Numeric.ftl:30]: Docs are based off of the java.lang.Math wording.

File: engine/function/src/templates/Numeric.ftl [hunk 5]
```diff
@@ -3378,6 +3390,656 @@ public class Numeric {
...
+     static public double expm1(${pt.primitive} x) {
+        if (isNull(x)) {
+            return NULL_DOUBLE;
+        }
+
+        return Math.expm1(x);
+     }
+
+    /**
+     * Returns the hypotenuse of a right-angled triangle, sqrt(x^2 +y^2), without intermediate overflow or underflow.
```
Comment (kosak) [inline engine/function/src/templates/Numeric.ftl:3470]: Space after +
```
sqrt(x^2 +y^2)
         ^
```

**Review by kosak — DISMISSED** [review]

**Review by kosak — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

Community: https://github.com/deephaven/deephaven-docs-community/issues/345

**Pull request merged**